rust/src/storage/file/mod.rs

```diff
@@ -58,25 +65,29 @@ impl StorageBackend for FileStorageBackend {
     }
 
     async fn put_obj(&self, path: &str, obj_bytes: &[u8]) -> Result<(), StorageError> {
-        let dir = std::path::Path::new(path);
+        let tmp_file_name = format!("{}.temporary", Uuid::new_v4().to_string());
+        let tmp_path = self.join_path(&self.root, &tmp_file_name);
 
-        if let Some(d) = dir.parent() {
-            fs::create_dir_all(d).await?;
+        // run this in loop in case tmp file with tmp_file_name already exists
+        loop {
+            let mut rf = fs::OpenOptions::new()
+                .create_new(true)
+                .write(true)
+                .open(&tmp_path)
```

Comment:
shouldn't we retry with new temp path here?

Reply:
yeah :)